### DISCUSSION TITLE
Possible colour/animation difference between stable and lazer hit/approach circles

### DISCUSSION THREAD

**Discussion by Theighlin** [ppy/osu#24872] [Q&A] [answered]
# Approach Rate "feels" different in lazer
After hearing from a couple players that AR felt different in lazer (as in harder), and reading it on [discord](https://discord.com/channels/188630481301012481/1097318920991559880/1153821955334623272) too earlier today, I decided to compare the 2 versions of the game to see what was different.

# Comparison Methodology
I recorded Lossless 144fps videos with OBS of both stable and lazer, making sure the variables matched between the two.

I made 2 comparison videos of the 2 games side by side (Second one after realising Combo Colour Normalisation was turned on).
I looked at the videos frame by frame, and the approach circles seemed to be timed correctly, as in they appeared and progressed on both version at the same time, so I tried to look for a difference in the colours, because i knew the brightness can make this feel different (players in stable sometimes turn up their gamma when playing very high ar, or use special skins).

Peppy suggested in the discord to try using a colour picker, but the colours tend to vary too much and i could never pick the same spot between the recordings or the games, best i could do was see if they were roughly the same in the same area.

https://github.com/ppy/osu/assets/53081750/a2ea9d7f-30ed-4c26-a052-043f8e055d18

# Conclusion
I opened the original videos, synced them and switched between the two.
It looks like lazer approach circles start out darker but end up brighter than stable. (maybe not colour but fade difference?)
It's possible that at high speeds this results in a shorter/faster feeling approach circle.

| Lazer | Stable |
|--------|--------|
| ![image](https://github.com/ppy/osu/assets/53081750/8279076c-4dbc-4bca-8a98-0b5d78937e53) | ![image](https://github.com/ppy/osu/assets/53081750/0c13271a-0c1a-4aed-a529-fb44df0dfd72) | 

It's also visible in this other example, where lazer is a fraction of a moment ahead in time, but still looks dimmer than stable
| Lazer | Stable |
|--------|--------|
| ![image](https://github.com/ppy/osu/assets/53081750/7c3ee61b-20f1-4ad3-a823-e1d3e406c09f ) | ![image](https://github.com/ppy/osu/assets/53081750/17fa7333-c49e-4695-b185-80bf122a0e89) |

The only setting i ever changed on this pc is "Digital Vibrance" in Nvidia Control Panel from 50% to 75% (didn't change since i started my investigation)
![image](https://github.com/ppy/osu/assets/53081750/16f505fd-54b5-46bf-8a34-a7fd3e6cd876)

# Files
I'm sharing the comparisons and original uncompressed videos in case they can be helpful https://drive.google.com/drive/folders/1SkiDmYT2KoqWgsDXw-miQj6HATISEzHr?usp=sharing.
Beatmap +DT: https://osu.ppy.sh/beatmapsets/1941215#osu/4014539
Skin: https://osu.ppy.sh/community/forums/topics/365036?n=1

I can also do more testing on different/higher AR maps or different skins if needed.

**Comment by peppy** [comment]
Good find here. The difference is that stable fades approach circles in to 90% opacity, while lazer fades to 100%.

I would like to think that this was a conscious decision by me in 2016 (https://github.com/ppy/osu/commit/3ac89216bd7ede6f34c403c795be50c81a37f782) but it was probably done blindly without checking stable code. Fading to 100% does feel more sane?

Anyways, question is what we want to do about this. I guess we should adjust to match stable and call it a day for now? cc @smoogipoo

**Reply by peppy** [reply, marked as answer]
Tracking at https://github.com/ppy/osu/issues/24941.

**Reply by smoogipoo** [reply]
The approach circle is one thing, but what about everything else being dimmer in lazer? I wouldn't mind matching approach circles fading to 90%, just making sure the rest isn't forgotten.

**Reply by peppy** [reply]
@smoogipoo see https://github.com/ppy/osu/discussions/24872#discussioncomment-7119652, directly below your comment.

**Comment by peppy** [comment]
Also touching on the "circles just appearing are dimmed more on stable" – that's note dim. We'd want to check why that doesn't match, but it's a separate issue to what is being tracked at #24941. You can tell because the dim is also applied to the number on the circle (and generally the circle itself, but probably isn't because of skins being skins here and skinning the wrong elements or something).

**Reply by peppy** [reply]
This is being tracked at https://github.com/ppy/osu/discussions/23571.

**Reply by peppy** [reply]
@Theighlin if you can run from source, could you try applying this diff and see if it helps? It's only a 1.4% difference so I'm not sure it's going to resolve the full issue, but who knows.

```diff
diff --git a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs
index 920dfcab03..f254b0bcaa 100644
--- a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs
@@ -8,6 +8,7 @@
 using osu.Framework.Bindables;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Primitives;
+using osu.Game.Graphics;
 using osu.Game.Rulesets.Judgements;
 using osu.Game.Rulesets.Objects.Drawables;
 using osu.Game.Rulesets.Osu.Judgements;
@@ -82,7 +83,11 @@ protected override void UpdateInitialTransforms()
                 // For now this is applied across all skins, and matches stable.
                 // For simplicity, dim colour is applied to the DrawableHitObject itself.
                 // We may need to make a nested container setup if this even causes a usage conflict (ie. with a mod).
-                this.FadeColour(new Color4(195, 195, 195, 255));
+
+                // In stable, the dim used differs per element.
+                // - Approach circles and hit circles (elements with combo colour applied) are 75% of combo colour.
+                // - Hit circle overlays (elements without combo colour applied) are Gray(195/255).
+                this.FadeColour(OsuColour.Gray(0.75f));
                 using (BeginDelayedSequence(InitialLifetimeOffset - OsuHitWindows.MISS_WINDOW))
                     this.FadeColour(Color4.White, 100);
             }

```

**Reply by Theighlin** [reply]
I'm not sure where this 1.4% difference i'm looking for is, final approach circle brightness? Anyway, i've looked at lazer with and without diff, i can't see any difference, or it is placebo if i do.

https://github.com/ppy/osu/assets/53081750/b3507928-8b88-4313-9542-358fe042188a

Note that while i can use the colour picker this way, I'm never sure the 2 frames are exactly the same, ideally, i would want to place both games at the same x milliseconds into the song, but with recording this is the best i can do.

**Comment by Theighlin** [comment]
For the record, there still seems to be a slight difference, but its visibility depends on the combo colour (managed to use colour picker somewhat accurately, most cases the hex value is the same). I'm seeing it mainly on white, but i also can't go around checking every colour imaginable. Not worth acting upon imo, maybe only if players still complain after next release, which i doubt.

| Lazer master | Stable |
|--------|--------|
| ![image](https://github.com/ppy/osu/assets/53081750/97f56936-e679-480b-999e-7b2f0cc841b2) | ![image](https://github.com/ppy/osu/assets/53081750/fec03085-f97d-473d-ba1f-6318d791482e) |

**Reply by peppy** [reply]
Have you turned off combo colour normalisation?

**Reply by Theighlin** [reply]
Yes, it was turned off.

**Reply by peppy** [reply]
Not too sure. Would probably need to see a few more examples of it happening across different combo colours to figure out what's going on. And whether it's worth actually fixing / altering to match.
